local-cli/runAndroid/runAndroid.js

```diff
@@ -78,13 +78,108 @@ function tryRunAdbReverse(device) {
 // Builds the app and runs it on a connected emulator / device.
 function buildAndRun(args) {
   process.chdir(path.join(args.root, 'android'));
+  const cmd = process.platform.startsWith('win')
+    ? 'gradlew.bat'
+    : './gradlew';
+
+  const packageName = fs.readFileSync(
+      'app/src/main/AndroidManifest.xml',
+      'utf8'
+    ).match(/package="(.+?)"/)[1];
+
+  const adbPath = getAdbPath();
+  if (args.deviceId) {
+    runOnSpecificDevice(args, cmd, packageName, adbPath);
+  } else {
+    runOnAllDevices(args, cmd, packageName, adbPath);
+  }
+}
+
+function runOnSpecificDevice(args, cmd, packageName, adbPath) {
+  let devices = adb.getDevices();
+  let selectedDevice = findMatchingDevice(devices, args.deviceId);
+  if (selectedDevice){
+    buildApk(cmd);
+    installAndLaunchOnDevice(args, selectedDevice, packageName, adbPath);
+  } else {
+      if (devices){
+        console.log('Could not find device with the id: "' + args.deviceId + '".');
+        console.log('Choose one of the following:');
+        child_process.execFileSync('adb', ['devices'], {
```

Comment:
Could you call `adb.getDevices` rather than reimplement it here?
https://github.com/facebook/react-native/blob/b9396cd74419dd4165fb051605076d19b1c18905/local-cli/runAndroid/adb.js#L38